common/logger/derive/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+    // Parse the input tokens into a syntax tree.
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = input.ident;
+    let fields = match input.data {
+        Data::Struct(DataStruct {
+            fields: Fields::Named(FieldsNamed { named, .. }),
+            ..
+        }) => named,
+        _ => panic!("derive(Schema) only supports structs with named fields"),
+    };
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+    let fields: Vec<StructField> = fields
+        .iter()
+        .map(|f| {
+            let ty = f.ty.clone();
+            let value_type = extract_attr(&f.attrs);
+
+            let inner_ty = extract_internal_type(&ty).cloned();
+            StructField {
+                value_type,
+                ident: f.ident.clone(),
+                ty: f.ty.clone(),
+                inner_ty,
+            }
+        })
+        .collect();
+
+    let setters = fields.iter().map(|f| {
+        let ident = &f.ident;
+
+        if let Some(ty) = &f.inner_ty {
+            quote! {
+                pub fn #ident(mut self, #ident: #ty) -> Self {
+                    self.#ident = ::std::option::Option::Some(#ident);
+                    self
+                }
+            }
+        } else {
+            let ty = &f.ty;
+            quote! {
+                pub fn #ident(mut self, #ident: #ty) -> Self {
+                    self.#ident = #ident;
+                    self
+                }
+            }
+        }
+    });
+
+    // Calls to visit_pair
+    let visitor = format_ident!("visitor");
+    let from_serde = quote! { ::libra_logger::Value::from_serde };
+    let from_display = quote! { ::libra_logger::Value::from_display };
+    let from_debug = quote! { ::libra_logger::Value::from_debug };
+    let key_new = quote! { ::libra_logger::Key::new };
+    let visits = fields.iter().map(|f| {
+        let ident = f.ident.as_ref().unwrap();
+        let ident_str = ident.to_string();
+
+        let from_fn = match f.value_type {
+            Some(ValueType::Display) => &from_display,
+            Some(ValueType::Debug) => &from_debug,
+            Some(ValueType::Serde) | None => &from_serde,
+        };
+        if f.inner_ty.is_some() {
+            quote! {
+                if let Some(#ident) = &self.#ident {
+                    #visitor.visit_pair(#key_new(#ident_str), #from_fn(#ident));
+                }
+            }
+        } else {
+            quote! {
+                #visitor.visit_pair(#key_new(#ident_str), #from_fn(&self.#ident));
+            }
+        }
+    });
+
+    // Build the output, possibly using quasi-quotation
+    let expanded = quote! {
+        impl #impl_generics #name #ty_generics #where_clause {
+            #(#setters)*
+
+            pub fn into_struct_log(self) -> ::libra_logger::StructuredLogEntry {
+                ::libra_logger::StructuredLogEntry::new_unnamed().schema(self)
+            }
```

Comment:
I guess we can't implement `into()` directly?

Reply:
I mean we probably can, but I figured we'd just be explicit about it right now. My goal is to eliminate this in the future so folks can just provide the schema and the macro can do the magic conversion itself.